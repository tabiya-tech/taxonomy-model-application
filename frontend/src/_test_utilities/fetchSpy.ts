export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}
