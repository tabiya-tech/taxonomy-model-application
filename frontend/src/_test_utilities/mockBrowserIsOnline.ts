import { fireEvent } from "@testing-library/react";

export function mockBrowserIsOnLine(isOnline: boolean) {
  if (window.navigator.onLine !== isOnline) {
    jest.spyOn(window.navigator, "onLine", "get").mockReturnValue(isOnline);
    if (isOnline) {
      fireEvent.online(window);
    } else {
      fireEvent.offline(window);
    }
  }
}

export function unmockBrowserIsOnLine() {
  jest.spyOn(window.navigator, "onLine", "get").mockRestore();
}
