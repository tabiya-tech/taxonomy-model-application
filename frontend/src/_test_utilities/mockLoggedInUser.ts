import { AuthContextValue, TabiyaUser, TabiyaUserRole, authContextDefaultValue } from "src/auth/AuthProvider";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>){
  let user: TabiyaUser = mockedValues.user || { username: "foo", roles: [] };
  jest.spyOn(require("src/auth/hooks/useAuthUser"), "useAuthUser").mockReturnValue({
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: TabiyaUserRole) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues
  });
}
