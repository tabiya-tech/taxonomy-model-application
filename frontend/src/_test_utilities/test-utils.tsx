// Based on https://testing-library.com/docs/react-testing-library/setup/

import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/providers";

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
