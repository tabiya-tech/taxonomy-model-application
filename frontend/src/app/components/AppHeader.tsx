import { Box, IconButton, Menu, MenuItem } from "@mui/material";
import { NavLink } from "react-router-dom";
import PermIdentityIcon from "@mui/icons-material/PermIdentity";
import { routerPaths } from "src/app/routerConfig";
import * as React from "react";
import { useState } from "react";
import { AuthContext, UserRole, UserRoleContextValue } from "../providers";

const uniqueId = "65b0785e-14d9-43a3-b260-869983312406";
export const DATA_TEST_ID = {
  APP_HEADER_CONTAINER: `app-header-container-${uniqueId}`,
  APP_HEADER_LOGO: `app-header-logo-${uniqueId}`,
  APP_HEADER_LOGO_LINK: `app-header-logo-link-${uniqueId}`,
  APP_HEADER_ICON_USER: `app-header-icon-user-${uniqueId}`,
};
const AppHeader = () => {
  const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null);
  const { userRole, setCookie } = React.useContext(AuthContext) as UserRoleContextValue;

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  const handleUser = () => {
    setCookie("authCookie", UserRole.ReadOnlyUser);
    setAnchorEl(null);
  };

  const handleModelManager = () => {
    setCookie("authCookie", UserRole.ModelManager);
    setAnchorEl(null);
  };

  const handleAdmin = () => {
    setCookie("authCookie", UserRole.Admin);
    setAnchorEl(null);
  };

  return (
    <Box
      display="flex"
      justifyContent="space-between"
      alignItems="center"
      data-testid={DATA_TEST_ID.APP_HEADER_CONTAINER}
    >
      <NavLink style={{ lineHeight: 0 }} to={routerPaths.ROOT} data-testid={DATA_TEST_ID.APP_HEADER_LOGO_LINK}>
        <img src="/logo.svg" alt="Tabiya" height={"30px"} data-testid={DATA_TEST_ID.APP_HEADER_LOGO} />
      </NavLink>
      <Box display="flex" alignItems="center" gap={2}>
        {userRole === UserRole.Admin ? "Admin" : userRole === UserRole.ReadOnlyUser ? "User" : "Model Manager"}
        <IconButton onClick={handleClick}>
          <PermIdentityIcon data-testid={DATA_TEST_ID.APP_HEADER_ICON_USER} />
        </IconButton>
        <Menu anchorEl={anchorEl} open={Boolean(anchorEl)} onClose={handleClose}>
          <MenuItem onClick={handleUser}>User</MenuItem>
          <MenuItem onClick={handleModelManager}>Model Manager</MenuItem>
          <MenuItem onClick={handleAdmin}>Admin</MenuItem>
        </Menu>
      </Box>
    </Box>
  );
};
export default AppHeader;
