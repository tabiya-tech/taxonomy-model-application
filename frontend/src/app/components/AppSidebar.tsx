import { NavLink } from 'react-router-dom';
import Box from '@mui/material/Box';
import FolderOutlinedIcon from '@mui/icons-material/FolderOutlined';
import SettingsIcon from '@mui/icons-material/Settings';
import PermIdentityIcon from '@mui/icons-material/PermIdentity';
import EditOutlinedIcon from '@mui/icons-material/EditOutlined';
import VerticalSplitOutlinedIcon from '@mui/icons-material/VerticalSplitOutlined';
import { styled } from '@mui/material';

const uniqueId = '579896ed-8a99-4a1e-8dc4-e4236d7b06df';
export const DATA_TEST_ID = {
  APP_SIDEBAR_CONTAINER: `app-sidebar-container-${uniqueId}`,
  APP_SIDEBAR_ICON_DIRECTORY: `app-sidebar-icon-directory-${uniqueId}`,
  APP_SIDEBAR_ICON_EXPLORE: `app-sidebar-icon-explore-${uniqueId}`,
  APP_SIDEBAR_ICON_EDIT: `app-sidebar-icon-edit-${uniqueId}`,
  APP_SIDEBAR_ICON_USER: `app-sidebar-icon-user-${uniqueId}`,
  APP_SIDEBAR_ICON_SETTINGS: `app-sidebar-icon-settings-${uniqueId}`,
  APP_SIDEBAR_DIRECTORY_LINK: `app-sidebar-directory-link-${uniqueId}`,
  APP_SIDEBAR_EXPLORE_LINK: `app-sidebar-explore-link-${uniqueId}`,
  APP_SIDEBAR_EDIT_LINK: `app-sidebar-edit-link-${uniqueId}`,
  APP_SIDEBAR_USER_LINK: `app-sidebar-user-link-${uniqueId}`,
  APP_SIDEBAR_SETTINGS_LINK: `app-sidebar-settings-link-${uniqueId}`,
};

export const ITEMS_LABEL_TEXT = {
  [DATA_TEST_ID.APP_SIDEBAR_ICON_DIRECTORY]: 'Directory',
  [DATA_TEST_ID.APP_SIDEBAR_ICON_EXPLORE]: 'Explore',
  [DATA_TEST_ID.APP_SIDEBAR_ICON_EDIT]: 'Model',
  [DATA_TEST_ID.APP_SIDEBAR_ICON_USER]: 'Users',
  [DATA_TEST_ID.APP_SIDEBAR_ICON_SETTINGS]: 'Settings',
};

const appSidebarItems = [
  {
    icon: (
      <FolderOutlinedIcon
        sx={{ width: '24px', height: '24px' }}
        data-testid={DATA_TEST_ID.APP_SIDEBAR_ICON_DIRECTORY}
        className='item-icon'
      />
    ),
    label: ITEMS_LABEL_TEXT[DATA_TEST_ID.APP_SIDEBAR_ICON_DIRECTORY],
    pathName: '/',
    iconLinkTestID: DATA_TEST_ID.APP_SIDEBAR_DIRECTORY_LINK,
  },
  {
    icon: (
      <VerticalSplitOutlinedIcon
        sx={{ width: '24px', height: '24px' }}
        data-testid={DATA_TEST_ID.APP_SIDEBAR_ICON_EXPLORE}
        className='item-icon'
      />
    ),
    label: ITEMS_LABEL_TEXT[DATA_TEST_ID.APP_SIDEBAR_ICON_EXPLORE],
    pathName: '/explore',
    iconLinkTestID: DATA_TEST_ID.APP_SIDEBAR_EXPLORE_LINK,
  },
  {
    icon: (
      <EditOutlinedIcon
        sx={{ width: '24px', height: '24px' }}
        data-testid={DATA_TEST_ID.APP_SIDEBAR_ICON_EDIT}
        className='item-icon'
      />
    ),
    label: ITEMS_LABEL_TEXT[DATA_TEST_ID.APP_SIDEBAR_ICON_EDIT],
    pathName: 'edit',
    iconLinkTestID: DATA_TEST_ID.APP_SIDEBAR_EDIT_LINK,
  },
  {
    icon: (
      <PermIdentityIcon
        sx={{ width: '24px', height: '24px' }}
        data-testid={DATA_TEST_ID.APP_SIDEBAR_ICON_USER}
        className='item-icon'
      />
    ),
    label: ITEMS_LABEL_TEXT[DATA_TEST_ID.APP_SIDEBAR_ICON_USER],
    pathName: '/users',
    iconLinkTestID: DATA_TEST_ID.APP_SIDEBAR_USER_LINK,
  },
  {
    icon: (
      <SettingsIcon
        sx={{ width: '24px', height: '24px' }}
        data-testid={DATA_TEST_ID.APP_SIDEBAR_ICON_SETTINGS}
        className='item-icon'
      />
    ),
    label: ITEMS_LABEL_TEXT[DATA_TEST_ID.APP_SIDEBAR_ICON_SETTINGS],
    pathName: '/settings',
    iconLinkTestID: DATA_TEST_ID.APP_SIDEBAR_SETTINGS_LINK,
  },
];

const CustomNavLink = styled(NavLink)`
  text-decoration: none;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 4px;
  color: #43474e;
  .item-icon {
    display: flex;
    align-items: center;
    padding: 4px 14px;
    border-radius: 16px;
  }
  &.active {
    .item-icon {
      background-color: #5cff9f;
    }
  }
  .item-label {
    font-size: 12px;
    color: #43474e;
  }
  &:hover {
    .item-icon {
      background-color: #5cff9f;
    }
  }
`;

const Container = styled(Box)`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 0 16px;
  //margin-top: 44px;
  font-family: ${({ theme }) => theme.typography.fontFamily};
  height: 100%;
  padding-left: 24px;
`;

const AppSidebar = () => {
  return (
    <Container
      gap='12px'
      data-testid={DATA_TEST_ID.APP_SIDEBAR_CONTAINER}
    >
      {appSidebarItems.map((item) => (
        <CustomNavLink key={item.pathName} to={item.pathName} data-testid={item.iconLinkTestID}>
          {item.icon}
          <span className='item-label'>{item.label}</span>
        </CustomNavLink>
      ))}
    </Container>
  );
};
export default AppSidebar;
