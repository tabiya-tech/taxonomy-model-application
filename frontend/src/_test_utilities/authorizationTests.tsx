import { render, screen } from "src/_test_utilities/test-utils";
import AuthAPISpecs from "api-specifications/auth";
import { mockLoggedInUser } from "./mockLoggedInUser";

export const ALL_USERS = [
  AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS,
  AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER,
  AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER,
];

type AuthorizationTestsParams = {
  name: string;
  Component: JSX.Element;
  testIds?: string[];
  roles?: AuthAPISpecs.Enums.TabiyaRoles[];
  tests?: Partial<
    Record<
      AuthAPISpecs.Enums.TabiyaRoles,
      {
        visible?: string[];
        hidden?: string[];
        visibleDisabled?: string[];
      }
    >
  >;
};

export const authorizationTests = {
  defaultName: "Test the rendering of the component based on Roles",
  callback: ({ name, Component, testIds = [], roles = ALL_USERS, tests }: AuthorizationTestsParams) => {
    return () => {
      test.each(roles?.map((role) => [role]))(`${name} renders correctly for user with role %s`, (role) => {
        const commonElements = testIds || [];
        const { visible = [], visibleDisabled = [], hidden = [] } = tests?.[role] || {};

        mockLoggedInUser({ user: { username: "foo", roles: [role] } });

        render(Component);

        commonElements.forEach((testId) => expect(screen.getByTestId(testId)).toBeInTheDocument());

        visible?.forEach((testId) => {
          expect(screen.getByTestId(testId)).toBeVisible();
        });

        hidden?.forEach((testId) => {
          expect(screen.queryByTestId(testId)).not.toBeInTheDocument();
        });

        visibleDisabled?.forEach((testId) => {
          const element = screen.getByTestId(testId);
          expect(element).toBeVisible();
          if (element instanceof HTMLInputElement) {
            expect(element).toBeDisabled();
          }
        });
      });
    };
  },
};
