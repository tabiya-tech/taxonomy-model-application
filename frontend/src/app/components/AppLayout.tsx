import React from 'react';
import Box from '@mui/material/Box';
import AppHeader from './AppHeader';
import AppSidebar from './AppSidebar';
import { styled, useTheme } from '@mui/material';


const Container = styled(Box)`
  display: flex;
  flex-direction: column;
  height: 100%;
  padding: ${({theme}) => theme.spacing(theme.tabiyaSpacing.lg)};
  padding-bottom: ${({theme}) => theme.spacing(theme.tabiyaSpacing.none)};
  gap: ${({theme}) => theme.spacing(theme.tabiyaSpacing.xl)};
`;

interface AppLayoutProps {
  children: React.ReactNode;
}

export const DATA_TEST_ID = {
  LAYOUT : "cabab611-7d62-417a-a4a7-5d8f7266ea92"
}

export const AppLayout: React.FC<AppLayoutProps> = ({ children }) => {
  const theme = useTheme()
  return (
    <Container
      data-testid={DATA_TEST_ID.LAYOUT}
      bgcolor='containerBackground.main'
    >
      <AppHeader />
      <Box display='flex' flexDirection='row' flex={1} gap={theme.tabiyaSpacing.xl} overflow='hidden'>
        <AppSidebar />
        <Box flex={1} display='flex'>
          {children}
        </Box>
      </Box>
    </Container>
  );
};

export default AppLayout;
