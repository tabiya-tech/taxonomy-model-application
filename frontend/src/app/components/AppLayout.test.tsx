import { render } from 'src/_test_utilities/test-utils';
import AppLayout, { DATA_TEST_ID } from './AppLayout';
import {DATA_TEST_ID as APP_SIDE_BAR_ID} from './AppSidebar';
import {DATA_TEST_ID as APP_HEADER_ID} from './AppHeader';

// Mocking the AppSidebar and AppHeader components
jest.mock('./AppSidebar', () => {
    const actual = jest.requireActual('./AppSidebar');
    const mockedAppSidebar = jest.fn().mockImplementation(
        () => <div data-testid={actual.DATA_TEST_ID.CONTAINER}>Mocked AppSidebar</div>)
    return {
      __esModule: true,
      ...actual,
      default: mockedAppSidebar,
    };
});

// Mocking the AppHeader component
jest.mock('./AppHeader', () => {
    const actual = jest.requireActual('./AppHeader');
    const mockedAppHeader = jest.fn().mockImplementation(
        () => <div data-testid={actual.DATA_TEST_ID.APP_HEADER_CONTAINER}>Mocked AppHeader</div>)

  return {
    __esModule: true,
    ...actual,
    default: mockedAppHeader
  };
});

describe('AppLayout Render', () => {
  test('should render app layout component', () => {
    // GIVEN a Layout Children component
    const LayoutChildren = () => <div>foo</div>

    // WHEN the AppLayout component is rendered
    const { getByTestId, queryByText } = render(<AppLayout><LayoutChildren/></AppLayout>);

    // THEN expect layout to be present in the document
    expect(getByTestId(DATA_TEST_ID.LAYOUT)).toBeInTheDocument();
    
    // AND expect the AppSidebar and AppHeader to be in the document
    expect(getByTestId(APP_SIDE_BAR_ID.CONTAINER)).toBeInTheDocument();
    expect(getByTestId(APP_HEADER_ID.APP_HEADER_CONTAINER)).toBeInTheDocument();

    // AND expect the child component to be in the document
    expect(queryByText('foo')).toBeInTheDocument();
  });
});
