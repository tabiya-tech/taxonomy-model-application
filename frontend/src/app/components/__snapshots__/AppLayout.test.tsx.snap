// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppLayout Render should render app layout component: cabab611-7d62-417a-a4a7-5d8f7266ea92 1`] = `
<div
  class="MuiBox-root css-1e3gqq1"
  data-testid="cabab611-7d62-417a-a4a7-5d8f7266ea92"
>
  <div
    data-testid="app-header-container-65b0785e-14d9-43a3-b260-869983312406"
  >
    Mocked AppHeader
  </div>
  <div
    class="MuiBox-root css-k287y7"
  >
    <div
      data-testid="app-sidebar-container-579896ed-8a99-4a1e-8dc4-e4236d7b06df"
    >
      Mocked AppSidebar
    </div>
    <div
      class="MuiBox-root css-1oz4csq"
    >
      <div>
        foo
      </div>
    </div>
  </div>
</div>
`;
